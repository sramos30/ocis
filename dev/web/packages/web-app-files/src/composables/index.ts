export * from './fileListHeaderPosition'
export * from './pagination'
export * from './resourcesViewDefaults'
export * from './router'
export * from './sort'

declare module 'vue/types/vue' {
  interface Vue {
    [key: string]: any
  }
}
