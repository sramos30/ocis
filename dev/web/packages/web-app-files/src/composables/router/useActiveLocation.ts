import { ref, Ref, watch } from '@vue/composition-api'
import { useRoute, useRouter } from 'web-pkg/src/composables'
import { ActiveRouteDirectorFunc } from '../../router'

/**
 * watches the current route and re-evaluates the provided active location director
 * on each route name update.
 *
 * @param director
 * @param comparatives
 */
export const useActiveLocation = <T extends string>(
  director: ActiveRouteDirectorFunc<T>,
  ...comparatives: T[]
): Ref<boolean> => {
  const value = ref(false)
  const router = useRouter()
  const currentRoute = useRoute()
  watch(
    currentRoute,
    () => {
      value.value = director(router, ...comparatives)
    },
    { immediate: true }
  )
  return value
}
