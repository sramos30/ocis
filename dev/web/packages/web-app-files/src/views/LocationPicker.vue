<template>
  <main id="files-location-picker" class="oc-flex oc-height-1-1">
    <div tabindex="-1" class="files-list-wrapper oc-width-expand">
      <div id="files-location-picker-app-bar" class="oc-p-s">
        <h1 class="location-picker-selection-info oc-mb" v-text="title" />
        <p
          class="oc-text-muted oc-text-meta"
          data-testid="location-picker-current-hint"
          v-text="currentHint"
        />
        <hr class="oc-mt-rm" />
        <oc-breadcrumb :items="breadcrumbs" class="oc-mb-s" />
        <oc-grid gutter="small" flex class="oc-flex-middle">
          <div>
            <oc-button
              id="location-picker-btn-cancel"
              size="small"
              @click="leaveLocationPicker(originalLocation)"
            >
              <translate>Cancel</translate>
            </oc-button>
          </div>
          <div>
            <oc-button
              id="location-picker-btn-confirm"
              variation="primary"
              appearance="filled"
              size="small"
              :disabled="!canConfirm"
              @click="confirmAction"
            >
              <span v-text="confirmBtnText" />
            </oc-button>
          </div>
        </oc-grid>
      </div>
      <div id="files-view">
        <app-loading-spinner v-if="loadResourcesTask.isRunning" />
        <template v-else>
          <no-content-message
            v-if="isEmpty"
            id="files-location-picker-empty"
            class="files-empty"
            icon="folder"
          >
            <template #message>
              <span v-translate>There are no resources in this folder.</span>
            </template>
          </no-content-message>
          <resource-table
            v-else
            id="files-location-picker-table"
            class="files-table"
            :are-thumbnails-displayed="false"
            :resources="paginatedResources"
            :disabled="disabledResources"
            :target-route="targetRoute"
            :has-actions="false"
            :is-selectable="false"
            :header-position="fileListHeaderY"
            :sort-by="sortBy"
            :sort-dir="sortDir"
            @sort="handleSort"
          >
            <template #footer>
              <pagination :pages="paginationPages" :current-page="paginationPage" />
              <list-info
                v-if="paginatedResources.length > 0"
                class="oc-width-1-1 oc-my-s"
                :files="totalFilesCount.files"
                :folders="totalFilesCount.folders"
                :size="totalFilesSize"
              />
            </template>
          </resource-table>
        </template>
      </div>
    </div>
  </main>
</template>

<script>
import { mapMutations, mapState, mapActions, mapGetters } from 'vuex'

import { basename, join } from 'path'
import ResourceTable from '../components/FilesList/ResourceTable.vue'
import { batchActions } from '../helpers/batchActions'
import { cloneStateObject } from '../helpers/store'
import MixinsGeneral from '../mixins'
import MixinFilesListFilter from '../mixins/filesListFilter'
import { useTask } from 'vue-concurrency'

import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import Pagination from '../components/FilesList/Pagination.vue'
import { DavProperties } from 'web-pkg/src/constants'
import { createLocationPublic, createLocationSpaces } from '../router'
import { buildWebDavFilesPath, buildWebDavSpacesPath } from '../helpers/resources'
import { useResourcesViewDefaults } from '../composables'
import { useCapabilitySpacesEnabled } from 'web-pkg/src/composables'

export default {
  metaInfo() {
    const title = `${this.title} - ${this.configuration.currentTheme.general.name}`

    return { title }
  },

  components: {
    ResourceTable,
    NoContentMessage,
    AppLoadingSpinner,
    ListInfo,
    Pagination
  },

  mixins: [MixinsGeneral, MixinFilesListFilter],

  setup() {
    const loadResourcesTask = useTask(function* (signal, ref, target) {
      ref.CLEAR_CURRENT_FILES_LIST()

      if (typeof target === 'object') {
        target = ref.target
      }

      const { context } = ref.$route.params

      let resources

      switch (context) {
        case 'public':
          resources = yield ref.$client.publicFiles.list(
            target,
            ref.publicLinkPassword,
            DavProperties.Default
          )
          break
        case 'space':
          resources = yield ref.$client.files.list(
            buildWebDavSpacesPath(ref.$route.query.storageId, target),
            1,
            DavProperties.Default
          )
          break
        default:
          resources = yield ref.$client.files.list(
            buildWebDavFilesPath(ref.user.id, target),
            1,
            DavProperties.Default
          )
      }

      ref.loadFiles({ currentFolder: resources[0], files: resources.slice(1) })
      ref.loadIndicators({
        client: ref.$client,
        currentFolder: ref.$route.params.item || '/'
      })
      ref.refreshFileListHeaderPosition()
    }).restartable()

    return {
      ...useResourcesViewDefaults({ loadResourcesTask }),
      hasSpaces: useCapabilitySpacesEnabled()
    }
  },

  data: () => ({
    originalLocation: ''
  }),

  computed: {
    ...mapState('Files', [
      'selectedResourcesForMove',
      'locationPickerTargetFolder',
      'currentFolder'
    ]),
    ...mapGetters('Files', ['publicLinkPassword', 'totalFilesCount', 'totalFilesSize']),
    ...mapGetters(['configuration']),
    ...mapGetters(['homeFolder']),
    ...mapState(['user']),

    title() {
      const personalRouteTitle = this.hasSpaces
        ? this.$gettext('Personal')
        : this.$gettext('All files')
      const target =
        basename(this.target) || personalRouteTitle
          ? this.$gettext('Personal')
          : this.$gettext('All files')
      const translated =
        this.currentAction === batchActions.move
          ? this.$gettext('Move into »%{ target }«')
          : this.$gettext('Copy into »%{ target }«')
      return this.$gettextInterpolate(translated, { target })
    },

    currentAction() {
      return this.$route.params.action
    },

    isPublicContext() {
      return this.$route.params.context === 'public'
    },

    target() {
      return this.$route.params.item || this.homeFolder
    },

    resources() {
      const resources = cloneStateObject(this.$route.query.resource)

      // In case there is only one resource, ensure that the return will still be an array
      if (typeof resources === 'string') {
        return [resources]
      }

      return resources
    },

    resourcesCount() {
      return this.resources.length
    },

    breadcrumbs() {
      const breadcrumbs = []
      const pathSegments = this.target.split('/').filter((item) => item !== '')

      if (this.isPublicContext) {
        const itemPath = join.apply(null, pathSegments.slice(0, 1))
        breadcrumbs.push(this.createBreadcrumbNode(0, this.$gettext('Public link'), itemPath))
        for (let i = 1; i < pathSegments.length; i++) {
          const itemPath = join.apply(null, pathSegments.slice(0, i + 1))
          breadcrumbs.push(this.createBreadcrumbNode(i + 1, pathSegments[i], itemPath))
        }
      } else {
        const personalRouteTitle = this.hasSpaces
          ? this.$gettext('Personal')
          : this.$gettext('All files')
        breadcrumbs.push(this.createBreadcrumbNode(0, personalRouteTitle, '/'))
        for (let i = 0; i < pathSegments.length; i++) {
          const itemPath = join.apply(null, pathSegments.slice(0, i + 1))
          breadcrumbs.push(this.createBreadcrumbNode(i + 1, pathSegments[i], itemPath))
        }
      }

      // the very last breadcrumb node is not supposed to be clickable
      delete breadcrumbs[breadcrumbs.length - 1].to

      return breadcrumbs
    },

    canConfirm() {
      return this.currentFolder && this.currentFolder.canCreate()
    },

    confirmBtnText() {
      switch (this.currentAction) {
        case batchActions.move:
          return this.$pgettext('Confirm action in the location picker for move', 'Move here')
        case batchActions.copy:
          return this.$pgettext('Confirm action in the location picker for copy', 'Paste here')
        default:
          return this.$gettext('Confirm')
      }
    },

    disabledResources() {
      const resources = cloneStateObject(this.paginatedResources)

      return resources
        .filter((resource) => resource.type !== 'folder' || this.resources.includes(resource.path))
        .map((resource) => resource.id)
    },

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    targetRoute() {
      const route = {
        name: this.$route.name,
        query: {
          resource: this.resources
        },
        params: {
          action: this.currentAction
        }
      }

      if (this.$route.query.storageId) {
        route.query.storageId = this.$route.query.storageId
      }

      return route
    },

    currentHint() {
      if (this.currentAction === 'move') {
        return this.$gettext(
          `Navigate into the desired folder and move selected resources into it.
          You can navigate into a folder by clicking on its name.
          To navigate back, you can click on the breadcrumbs.
          Resources will be moved into the folder where you are currently located.`
        )
      }

      return this.$gettext(
        `Navigate into the desired folder and copy selected resources into it.
        You can navigate into a folder by clicking on its name.
        To navigate back, you can click on the breadcrumbs.
        Resources will be copied into the folder where you are currently located.`
      )
    }
  },

  watch: {
    $route: {
      handler: function (to, from) {
        const sameRoute = to.name === from?.name
        const sameItem = to.params?.item === from?.params?.item
        if (!sameRoute || !sameItem) {
          this.loadResourcesTask.perform(this, this.$route)
        }
      },
      immediate: true
    }
  },

  created() {
    this.originalLocation = this.target
  },

  methods: {
    ...mapMutations('Files', ['CLEAR_CURRENT_FILES_LIST']),
    ...mapActions('Files', ['loadFiles', 'loadIndicators']),
    ...mapActions(['showMessage']),

    createBreadcrumbNode(index, text, itemPath) {
      return {
        index,
        text,
        to: {
          name: this.$route.name,
          params: {
            action: this.currentAction,
            item: itemPath
          },
          query: {
            resource: this.resources
          }
        }
      }
    },

    leaveLocationPicker(target) {
      switch (this.$route.params.context) {
        case 'public':
          this.$router.push(
            createLocationPublic('files-public-files', { params: { item: target } })
          )
          break
        case 'space':
          this.$router.push(
            createLocationSpaces('files-spaces-project', {
              params: { storageId: this.$route.query.storageId, item: target || '/' }
            })
          )
          break
        default:
          this.$router.push(
            createLocationSpaces('files-spaces-personal-home', { params: { item: target || '/' } })
          )
      }
    },

    isRowDisabled(resource) {
      if (resource.type !== 'folder' || !resource.canCreate()) {
        return true
      }
      return this.resources.some((item) => item === resource.path)
    },

    async confirmAction() {
      const errors = []
      let promise = null

      // Execute move or copy
      for (const resource of this.resources) {
        let targetPath
        let resourceName = basename(resource)
        const exists = this.paginatedResources.some((item) => {
          return basename(item.name) === resourceName
        })
        if (exists) {
          const message = this.$gettext('Resource with name %{name} already exists')
          errors.push({
            resource: resourceName,
            message: this.$gettextInterpolate(message, { name: resourceName }, true)
          })
          continue
        }

        switch (this.currentAction) {
          case batchActions.move: {
            switch (this.$route.params.context) {
              case 'public':
                targetPath = `${this.target || '/'}`
                targetPath += `/${resourceName}`
                promise = this.$client.publicFiles.move(
                  resource,
                  targetPath,
                  this.publicLinkPassword
                )
                break
              case 'space':
                targetPath = buildWebDavSpacesPath(this.$route.query.storageId, this.target || '/')
                targetPath += `/${resourceName}`
                promise = this.$client.files.move(
                  buildWebDavSpacesPath(this.$route.query.storageId, resource),
                  targetPath
                )
                break
              default:
                targetPath = buildWebDavFilesPath(this.user.id, this.target || '/')
                targetPath += `/${resourceName}`
                promise = this.$client.files.move(
                  buildWebDavFilesPath(this.user.id, resource),
                  targetPath
                )
            }
            break
          }
          case batchActions.copy: {
            switch (this.$route.params.context) {
              case 'public':
                targetPath = `${this.target || '/'}`
                targetPath += `/${resourceName}`

                promise = this.$client.publicFiles.copy(
                  resource,
                  targetPath,
                  this.publicLinkPassword
                )
                break
              case 'space':
                targetPath = buildWebDavSpacesPath(this.$route.query.storageId, this.target || '/')
                resourceName = basename(resource)
                targetPath += `/${resourceName}`
                promise = this.$client.files.copy(
                  buildWebDavSpacesPath(this.$route.query.storageId, resource),
                  targetPath
                )
                break
              default:
                targetPath = buildWebDavFilesPath(this.user.id, this.target || '/')
                targetPath += `/${resourceName}`
                promise = this.$client.files.copy(
                  buildWebDavFilesPath(this.user.id, resource),
                  targetPath
                )
            }
            break
          }
          default:
            return
        }

        await promise.catch((error) => {
          error.resource = resourceName
          errors.push(error)
        })
      }

      // Leave location picker if everything was successful
      if (errors.length === 0) {
        this.leaveLocationPicker(this.target)
        return
      }

      // Display error messages
      let title = ''
      if (errors.length === 1) {
        switch (this.currentAction) {
          case batchActions.move: {
            title = this.$gettext('Failed to move "%{resourceName}"')
            break
          }
          case batchActions.copy: {
            title = this.$gettext('Failed to copy "%{resourceName}"')
            break
          }
          default:
            return
        }

        this.showMessage({
          title: this.$gettextInterpolate(title, { resourceName: errors[0].resource }, true),
          status: 'danger'
        })

        return
      }

      switch (this.currentAction) {
        case batchActions.move: {
          title = this.$gettext('Failed to move %{count} resources')
          break
        }
        case batchActions.copy: {
          title = this.$gettext('Failed to copy %{count} resources')
          break
        }
        default:
          return
      }

      this.showMessage({
        title: this.$gettextInterpolate(title, { count: errors.length }),
        status: 'danger'
      })
    }
  }
}
</script>

<style lang="scss" scoped>
#files-location-picker {
  height: 100%;
  max-height: 100%;
  overflow-y: hidden;

  &-app-bar {
    position: sticky;
    top: 0;
    height: auto;
    z-index: 1;
    grid-area: header;
    background-color: var(--oc-color-background-default);
    box-sizing: border-box;
  }
}

.files-list-wrapper {
  height: 100%;
  max-height: 100%;
  overflow-y: auto;
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: max-content 1fr;
  gap: 0 0;
  grid-template-areas:
    'header'
    'main';

  &:focus {
    outline: none;
  }
}

#files-view {
  grid-area: main;
}

#files-location-picker-table ::deep tr {
  td:first-of-type {
    padding-left: 30px;
  }
}
</style>
