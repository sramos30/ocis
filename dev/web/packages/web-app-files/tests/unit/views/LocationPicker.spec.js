import { createFile, getStore, localVue } from './views.setup.js'
import LocationPicker from '@files/src/views/LocationPicker.vue'
import { mount, RouterLinkStub, shallowMount } from '@vue/test-utils'
import { createLocationOperations } from '../../../src/router'
import omit from 'lodash-es/omit'

const component = { ...LocationPicker, mounted: jest.fn() }

const defaultStubs = {
  'oc-breadcrumb': true,
  'oc-grid': true,
  'oc-button': true,
  'oc-icon': true,
  'oc-table-files': true
}

localVue.prototype.$client.files = {
  list: async () => [],
  move: async () => {}
}
localVue.prototype.$client.publicFiles = {
  list: async () => [],
  move: async () => {}
}

const router = {
  push: jest.fn(),
  afterEach: jest.fn(),
  currentRoute: {
    ...createLocationOperations('files-operations-location-picker'),
    query: {}
  },
  resolve: (r) => {
    return { href: r.name }
  }
}

const listLoaderStub = 'app-loading-spinner-stub'
const breadcrumbStub = 'oc-breadcrumb-stub'
const listInfoStub = 'list-info-stub'
const filesPaginationStub = 'pagination-stub'
const translateStub = 'translate-stub'

const selectors = {
  selectionInfo: '.location-picker-selection-info',
  currentHint: '[data-testid="location-picker-current-hint"]',
  cancelButton: '#location-picker-btn-cancel',
  confirmButton: '#location-picker-btn-confirm',
  filesView: '#files-view',
  noContentMessage: '#files-location-picker-empty',
  filesTable: '#files-location-picker-table',
  paginationCurrentItem: '.oc-pagination-list-item-current'
}

describe('LocationPicker', () => {
  describe('files app bar', () => {
    it.each([
      { action: 'move', item: '' },
      { action: 'copy', item: '' },
      { action: 'move', item: '/parent/child' },
      { action: 'move', item: '/parent' },
      { action: 'copy', item: '/parent/child' },
      { action: 'copy', item: '/parent' },
      { action: 'random-action', item: '/सिम्पल फोल्डर/टेश्ट' },
      { action: '', item: '/parent/child' },
      { action: true, item: '/parent/child' },
      { action: {}, item: '/parent/child' },
      {
        action: 'random-action',
        item: '/सिम्पल फोल्डर/टेश्ट फोल्डर'
      },
      {
        action: 'random-action',
        item: '/123/456'
      },
      {
        action: 'random-action',
        item: '/😂/🔥'
      }
    ])('should show location picker selection info according to route action and item', (input) => {
      const wrapper = getShallowWrapper({
        $route: getRoute({ action: input.action, item: input.item })
      })

      const selection = wrapper.find(selectors.selectionInfo)

      expect(selection.exists()).toBeTruthy()
      expect(selection).toMatchSnapshot()
    })
    it.each(['copy', 'move'])(
      'should show current hint according to the current route action',
      (action) => {
        const wrapper = getShallowWrapper({ $route: getRoute({ action: action }) })
        const currentHint = wrapper.find(selectors.currentHint)

        expect(currentHint.exists()).toBeTruthy()
        expect(currentHint).toMatchSnapshot()
      }
    )
    describe('should display breadcrumb with items', () => {
      it.each(['parent/child', 'parent//child', '/parent/child', '/parent/child/'])(
        'if route params context is public',
        (input) => {
          const route = getRoute({ context: 'public', item: input })
          const wrapper = getShallowWrapper({ $route: route })
          const breadcrumbItems = wrapper.find(breadcrumbStub).props().items

          expect(breadcrumbItems.length).toBe(2)
          expect(breadcrumbItems[0]).toMatchObject({
            index: 0,
            text: 'Public link',
            to: {
              name: route.name,
              params: {
                action: route.params.action,
                item: 'parent'
              },
              query: {
                resource: [route.query.resource]
              }
            }
          })
          expect(breadcrumbItems[1]).toMatchObject({
            index: 2,
            text: 'child'
          })
          expect(breadcrumbItems[1].to).toBeUndefined()
        }
      )
      it.each(['parent/child', 'parent//child', '/parent/child', '/parent/child/'])(
        'if route params context is not public',
        (input) => {
          const route = getRoute({ item: input })
          const wrapper = getShallowWrapper({ $route: route })
          const breadcrumbItems = wrapper.find(breadcrumbStub).props().items

          expect(breadcrumbItems.length).toBe(3)
          expect(omit(breadcrumbItems[0], ['text'])).toMatchObject({
            index: 0,
            to: {
              name: route.name,
              params: { action: route.params.action, item: '/' },
              query: {
                resource: [route.query.resource]
              }
            }
          })
          expect(breadcrumbItems[1]).toMatchObject({
            index: 1,
            text: 'parent',
            to: {
              name: route.name,
              params: { action: route.params.action, item: 'parent' },
              query: {
                resource: [route.query.resource]
              }
            }
          })
          expect(breadcrumbItems[2]).toMatchObject({
            index: 2,
            text: 'child'
          })
          expect(breadcrumbItems[2].to).toBeUndefined()
        }
      )
    })
    describe('cancel button', () => {
      it('should show the location picker cancel button', () => {
        const wrapper = getShallowWrapper()
        const cancelButton = wrapper.find(selectors.cancelButton)

        expect(cancelButton.exists()).toBeTruthy()
        expect(cancelButton.find(translateStub).exists()).toBeTruthy()
      })
      it('should call "leaveLocationPicker" method when clicked', async () => {
        const spyLeaveLocationPicker = jest.spyOn(LocationPicker.methods, 'leaveLocationPicker')
        const spyRouterPush = jest.spyOn(router, 'push')

        const wrapper = getMountedWrapper()
        const cancelButton = wrapper.find(selectors.cancelButton)

        await cancelButton.trigger('click')

        expect(spyLeaveLocationPicker).toHaveBeenCalledTimes(1)
        expect(spyLeaveLocationPicker).toHaveBeenCalledWith('/some/item')
        expect(spyRouterPush).toHaveBeenCalledTimes(1)
        expect(spyRouterPush).toHaveBeenCalledWith({
          name: 'files-spaces-personal-home',
          params: {
            item: '/some/item',
            storage: 'home'
          }
        })
      })
    })
    describe('should show the location picker confirm button', () => {
      it('should be disabled if no current folder', () => {
        const wrapper = getShallowWrapper()

        const confirmButton = wrapper.find(selectors.confirmButton)

        expect(confirmButton.exists()).toBeTruthy()
        expect(confirmButton.attributes().disabled).toBe('true')
      })
      it('should be disabled if current folder does not have "canCreate" property', () => {
        const route = getRoute()
        const store = createStore({ currentFolder: { canCreate: jest.fn(() => false) } })
        const wrapper = getShallowWrapper({ store: store, $route: route })

        const confirmButton = wrapper.find(selectors.confirmButton)

        expect(confirmButton.exists()).toBeTruthy()
        expect(confirmButton.attributes().disabled).toBe('true')
      })
      it.each([
        { action: 'move', expectedText: 'Move here' },
        { action: 'copy', expectedText: 'Paste here' },
        { action: 'anything else', expectedText: 'Confirm' }
      ])('should set button text according to current action', (input) => {
        const wrapper = getShallowWrapper({
          store: createStore({ currentFolder: { canCreate: jest.fn(() => true) } }),
          $route: getRoute({ action: input.action })
        })

        const confirmButton = wrapper.find(selectors.confirmButton)

        expect(confirmButton.exists()).toBeTruthy()
        expect(confirmButton.attributes().disabled).toBeUndefined()
        expect(confirmButton.text()).toBe(input.expectedText)
      })
      it('should call "confirmAction" method when clicked', async () => {
        const spyConfirmAction = jest.spyOn(LocationPicker.methods, 'confirmAction')
        const wrapper = getMountedWrapper({
          store: createStore({ currentFolder: { canCreate: jest.fn(() => true) } })
        })
        const confirmButton = wrapper.find(selectors.confirmButton)

        await confirmButton.trigger('click')
        expect(spyConfirmAction).toHaveBeenCalledTimes(1)
      })
    })
  })

  describe('files view', () => {
    describe('when the view is still loading', () => {
      let filesView
      beforeEach(() => {
        const wrapper = getShallowWrapper({ loading: true })
        filesView = wrapper.find(selectors.filesView)
      })
      it('should show app-loading-spinner component', () => {
        expect(filesView.find(listLoaderStub).exists()).toBeTruthy()
        expect(filesView.find(listLoaderStub)).toMatchSnapshot()
      })
      it('should not show no content message', () => {
        expect(filesView.find(selectors.noContentMessage).exists()).toBeFalsy()
      })
      it('should not show no location picker table', () => {
        expect(filesView.find(selectors.filesTable).exists()).toBeFalsy()
      })
    })
    describe('when the view is not loading anymore', () => {
      const wrapper = getShallowWrapper()
      const filesView = wrapper.find(selectors.filesView)
      it('should not show app-loading-spinner component', () => {
        expect(filesView.find(listLoaderStub).exists()).toBeFalsy()
      })
      it('should show no content message if active files length is less than one', () => {
        expect(filesView.find(selectors.noContentMessage).exists()).toBeTruthy()
        expect(filesView.find(selectors.noContentMessage)).toMatchSnapshot()
      })
      describe('when active files length is one or more', () => {
        const expectedResources = [
          { id: 100, name: 'file0.txt', type: 'file', path: '/home/file0.txt' }
        ]
        const store = createStore({
          totalFilesCount: { files: 2, folders: 0 },
          totalFilesSize: 0
        })
        const route = getRoute({ resource: '/home/file0.txt' })
        const wrapper = getShallowWrapper({
          store: store,
          $route: route,
          setup: {
            paginatedResources: expectedResources
          }
        })
        const filesView = wrapper.find(selectors.filesView)

        it('should not show the no content message component', () => {
          expect(filesView.find(selectors.noContentMessage).exists()).toBeFalsy()
        })
        it('should show location picker table', () => {
          expect(filesView.find(selectors.filesTable).exists()).toBeTruthy()

          const actualProps = filesView.find(selectors.filesTable).props()
          expect(actualProps.resources).toMatchObject(expectedResources)
          expect(actualProps.targetRoute).toMatchObject({
            name: route.name,
            query: {
              resource: [route.query.resource]
            },
            params: { action: route.params.action }
          })
          expect(actualProps.disabled).toMatchObject([expectedResources[0].id])
        })
        it.each([
          {
            id: 100,
            name: 'file0.txt',
            type: 'file',
            path: '/home/file0.txt',
            expectedState: [100]
          },
          {
            id: 101,
            name: 'some-resource',
            type: 'file',
            path: '/home/some-resource',
            expectedState: [101]
          },
          {
            id: 102,
            name: 'some-resource',
            type: 'folder',
            path: '/home/some-resource',
            expectedState: [102]
          },
          {
            id: 103,
            name: 'simple-folder',
            type: 'folder',
            path: '/simple-folder',
            expectedState: []
          }
        ])(
          'should set location item as disabled if not a folder or if its path is included in the active resource',
          (input) => {
            const route = getRoute({ resource: '/home/some-resource' })
            const wrapper = getShallowWrapper({
              data: { loading: false },
              $route: route,
              setup: {
                paginatedResources: [input]
              }
            })

            const filesView = wrapper.find(selectors.filesView)
            expect(filesView.find(selectors.filesTable).props().disabled).toMatchObject(
              input.expectedState
            )
          }
        )

        it('should show pagination if there is one or more pages', () => {
          const expectedResources = [createFile({ id: 100 }), createFile({ id: 101 })]
          const store = createStore({
            totalFilesCount: { files: 1000, folders: 1000 },
            totalFilesSize: 100000
          })
          const route = getRoute({ resource: '/home/some-resource' })
          const setup = {
            paginatedResources: expectedResources,
            paginationPages: 2,
            paginationPage: 1
          }
          const wrapper = getMountedWrapper({
            data: { loading: false },
            store: store,
            $route: route,
            setup: setup
          })
          const filesView = wrapper.find(selectors.filesView)
          const tablePagination = filesView.find(filesPaginationStub)

          expect(tablePagination.exists()).toBeTruthy()
          expect(tablePagination.props()).toMatchObject({
            pages: setup.paginationPages,
            currentPage: setup.paginationPage
          })
        })

        it('should load list info component with correct prop values', () => {
          const expectedResources = [createFile({ id: 100 }), createFile({ id: 101 })]
          const store = createStore({
            totalFilesCount: { files: 1, folders: 1 },
            totalFilesSize: 100
          })
          const route = getRoute({ resource: '/home/some-resource' })
          const wrapper = getMountedWrapper({
            data: { loading: false },
            store: store,
            $route: route,
            setup: {
              paginatedResources: expectedResources
            }
          })
          const filesView = wrapper.find(selectors.filesView)
          const listInfo = filesView.find(listInfoStub)

          expect(listInfo.exists()).toBeTruthy()
          expect(listInfo.props()).toMatchObject({
            files: store.getters['Files/totalFilesCount'].files,
            folders: store.getters['Files/totalFilesCount'].folders,
            size: store.getters['Files/totalFilesSize']
          })
        })
      })
    })
  })

  function getRoute({
    name = 'some-name',
    path = '/some/path',
    action = 'move',
    context = null,
    item = '/some/item',
    resource = 'some-resource'
  } = {}) {
    return {
      name: name,
      path: path,
      params: {
        action: action,
        item: item,
        context: context
      },
      query: {
        resource: resource
      }
    }
  }

  function createStore({
    publicLinkPassword = null,
    currentFolder = null,
    totalFilesCount = { files: null, folders: null },
    totalFilesSize = null,
    generalThemeName = ''
  } = {}) {
    return getStore({
      publicLinkPassword: publicLinkPassword,
      currentFolder: currentFolder,
      totalFilesSize: totalFilesSize,
      totalFilesCount: totalFilesCount,
      generalThemeName: generalThemeName,
      user: { id: 'test' }
    })
  }

  function mountOptions(store, $route, loading, setup = {}, stubs = defaultStubs) {
    return {
      localVue,
      store: store,
      stubs,
      mocks: {
        $route,
        $router: router
      },
      setup: () => ({
        loadResourcesTask: {
          isRunning: loading,
          perform: jest.fn()
        },
        ...setup
      })
    }
  }

  function getShallowWrapper({
    store = createStore(),
    $route = getRoute(),
    loading = false,
    setup
  } = {}) {
    return shallowMount(component, mountOptions(store, $route, loading, setup))
  }

  function getMountedWrapper({
    store = createStore(),
    $route = getRoute(),
    loading = false,
    setup
  } = {}) {
    return mount(
      component,
      mountOptions(store, $route, loading, setup, {
        'oc-button': false,
        'list-info': true,
        translate: true,
        pagination: true,
        RouterLink: RouterLinkStub
      })
    )
  }
})
