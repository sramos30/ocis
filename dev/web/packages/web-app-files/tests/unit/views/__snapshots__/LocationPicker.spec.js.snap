// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationPicker files app bar should show current hint according to the current route action 1`] = `
<p data-testid="location-picker-current-hint" class="oc-text-muted oc-text-meta">Navigate into the desired folder and copy selected resources into it.
  You can navigate into a folder by clicking on its name.
  To navigate back, you can click on the breadcrumbs.
  Resources will be copied into the folder where you are currently located.</p>
`;

exports[`LocationPicker files app bar should show current hint according to the current route action 2`] = `
<p data-testid="location-picker-current-hint" class="oc-text-muted oc-text-meta">Navigate into the desired folder and move selected resources into it.
  You can navigate into a folder by clicking on its name.
  To navigate back, you can click on the breadcrumbs.
  Resources will be moved into the folder where you are currently located.</p>
`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 1`] = `<h1 class="location-picker-selection-info oc-mb">Move into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 2`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 3`] = `<h1 class="location-picker-selection-info oc-mb">Move into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 4`] = `<h1 class="location-picker-selection-info oc-mb">Move into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 5`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 6`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 7`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 8`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 9`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 10`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 11`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 12`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files app bar should show location picker selection info according to route action and item 13`] = `<h1 class="location-picker-selection-info oc-mb">Copy into »Personal«</h1>`;

exports[`LocationPicker files view when the view is not loading anymore should show no content message if active files length is less than one 1`] = `<no-content-message-stub icon="folder" id="files-location-picker-empty" class="files-empty"></no-content-message-stub>`;

exports[`LocationPicker files view when the view is still loading should show app-loading-spinner component 1`] = `<app-loading-spinner-stub></app-loading-spinner-stub>`;
